### PR TITLE
[SPARK-24858][SQL] Avoid unnecessary parquet footer reads

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Currently the same Parquet footer is read twice in the function `buildReaderWithPartitionValues` of ParquetFileFormat if filter push down is enabled.

Fix it with simple changes.
## How was this patch tested?

Unit test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 1]
```diff
@@ -384,12 +385,10 @@ class ParquetFileFormat
       // *only* if the file was created by something other than "parquet-mr", so check the actual
       // writer here for this file.  We have to do this per-file, as each file in the table may
       // have different writers.
-      def isCreatedByParquetMr(): Boolean = {
-        val footer = ParquetFileReader.readFooter(sharedConf, filePath, SKIP_ROW_GROUPS)
-        footer.getFileMetaData().getCreatedBy().startsWith("parquet-mr")
-      }
+      val isCreatedByParquetMr = footerFileMetaData.getCreatedBy().startsWith("parquet-mr")
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Is `lazy` better? `timestampConversion` default is false.
```scala
lazy val isCreatedByParquetMr = footerFileMetaData.getCreatedBy().startsWith("parquet-mr")
```

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Yes, I just push the code to make it lazy value.

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: I think making `footerFileMetaData` lazy value is better.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Hm? `isCreatedByParquetMr` will be evaluated here. We should make `isCreatedByParquetMr` lazy too ..

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Also, let's leave a comment here saying that we avoid reading it by short circuit. I already saw multiple people confused here.

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

**Pull request closed without merging**